src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts

```diff
@@ -0,0 +1,131 @@
+        const sumPositive = this.sumCredits(positive);
+        const sumNegative = this.sumCredits(negative);
+
+        if (sumPositive + sumNegative < 0) {
+            return this.clearCredits(feedbackNodes);
+        }
+
+        const positiveSubtracted = this.subtractCredits(positive, sumNegative);
+
+        return [...positiveSubtracted, ...neutral, ...this.absCredits(negative)];
+    };
+
+    private subtractCredits = (feedbackNodes: FeedbackNode[], subtrahend: number): FeedbackNode[] => {
+        return feedbackNodes.map((node) => {
+            let credits = 0;
+            const current = node.credits ?? 0;
+            if (current + subtrahend >= 0) {
+                credits = current + subtrahend;
+            }
+
+            subtrahend = Math.min(subtrahend + current, 0);
+
+            return {
+                ...node,
+                credits,
+            };
+        });
+    };
+    /*
+     * Separates a list of feedback nodes by node credits. Has runtime of O(3n)
+     * @param feedbackNodes
+     * @return Tuple with values [Positive, Neutral, Negative]
+     */
+    private separateByCredits = (feedbackNodes: FeedbackNode[]): [FeedbackNode[], FeedbackNode[], FeedbackNode[]] => {
+        return [
+            feedbackNodes.filter((node) => (node.credits ?? 0) > 0),
+            feedbackNodes.filter((node) => (node.credits ?? 0) === 0),
+            feedbackNodes.filter((node) => (node.credits ?? 0) < 0),
+        ];
+    };
+
+    private sumCredits = (feedbackNodes: FeedbackNode[]) => {
+        return feedbackNodes.reduce((acc, node) => (node.credits ?? 0) + acc, 0);
+    };
+
+    /*
+     * Sets credits in nodes to absolute value
+     */
+    private absCredits = (feedbackNodes: FeedbackNode[]) => {
+        return feedbackNodes.map((node) => {
+            node.credits = Math.abs(node.credits ?? 0);
+            return node;
+        });
+    };
+
+    /*
+     * Sets credits to 0 for all feedback nodes
+     */
+    private clearCredits = (feedbackNodes: FeedbackNode[]) => {
+        return feedbackNodes.map((node) => ({
+            ...node,
+            credits: 0,
+        }));
+    };
+
+    private roundToDecimals = (i: number, n: number) => {
+        const f = 10 ** n;
+        return round(i, f);
+    };
+
+    private capCredits = (credits: number, maxCredits?: number): number => {
+        // no maxCredits or credits and maxCredits do not have the same sign;
+        if (!maxCredits || credits * maxCredits < 0) {
+            return credits;
+        }
+
+        const absCredits = Math.abs(credits);
+        const absMaxCredits = Math.abs(maxCredits);
+        return Math.sign(credits) * Math.min(absCredits, absMaxCredits);
+    };
+
+    private calculatePercentage = (node: FeedbackNode, maxPoints: number) => {
+        const appliedCredits = this.capCredits(node.credits ?? 0, node.maxCredits);
+        return this.roundToDecimals((appliedCredits / maxPoints) * 100, 2);
```

Comment:
We have a rounding function in utils.ts to adhere to the specified accuracy in the course

Reply:
Will look into that thanks!